spark/src/main/java/org/apache/iceberg/spark/source/Reader.java

```diff
@@ -308,6 +332,29 @@ private ReadTask(CombinedScanTask task, String tableSchemaString, String expecte
         encryptionManager.value(), caseSensitive);
     }
 
+    @Override
+    public String[] preferredLocations() {
```

Comment:
In other formats, splits typically fetch block locations as they were accessed. I think there is an assumption that split planning happens in parallel and that working with splits is cheap after that point.

My concern is that this defers fetching block locations until they are accessed, which will make a previously cheap call (retrieve locality from an instance field) into an expensive RPC call (ask the NN). I think it may be better to fetch this information when the `ReadTask` is created into a `transient` field. That way, the expensive part is done in parallel during split planning.